Hybridized finite elements

lindsayad opened the pull request:
Todo:

- [x] Documentation
- [x] framework level test

Closes #26405 #26406

lindsayad committed:
Create HybridizedKernel interface

Refs #26406

lindsayad committed:
Remove unused bnd_id from elem and neighbor face reinit APIs

Refs #26406

lindsayad committed:
Factor out a static prepareFace routine in NonlinearThread

I'll be using this routine in the hybridized kernels hopefully

Refs #26406

lindsayad committed:
Developing hybridized framework

- Update `HybridizedKernel` interface so that users only implement
  `assemble`. We can add more interfaces as required
- Create `HybridizedKernel::lm_increment_vector_name` data member
- Create and populate the ghosted LM increment vector
- Move code common to `NonlinearThread` and `PreCheckThread` down
  into `ThreadedElementLoop`
- Have `NonlinearSystem` inherit from `NonlinearImplicitSystem::ComputePreCheck`
  and implement the `precheck` function which we use to call the
  `PreCheckThread` which will compute the primal solution increment
  based on the lm solution increment which we compute from the
  `search_direction` given to us by PETSc

Refs #26406

lindsayad committed:
More fleshing out of HybridizedKernel interface

Refs #26406

lindsayad committed:
Create HybridizedIntegratedBC

Refs #26406

lindsayad committed:
Create hybridized NS objects

Refs #26405

lindsayad committed:
Some vector finite element related fixes

- Use more general `FEInterface::field_type` instead of manually listing
  finite element families that quickly become out-of-date
- Add more fine-gained control for requesting and retrieving divergence
  of shape functions

lindsayad committed:
Hybridization fixes

- Do allow call to `computeResidual`
- Register base as `Kernel`
- Always compute residual and Jacobian together

Refs #26406

lindsayad committed:
Working MOOSE HDG-NS lid driven input

Refs #26405

lindsayad committed:
Create MMS cavity test

Refs #26405

lindsayad committed:
Test both hierarchics and Lagrange

Refs #26405

lindsayad committed:
Allow L2_LAGRANGE_VEC variables

Refs #26406

lindsayad committed:
Create channel mms tests

Refs #26405

lindsayad committed:
Add physical lid and channel flow tests

Refs #26405

lindsayad committed:
Shorten mms tests by reducing refinement

Refs #26405

lindsayad committed:
Cut MMS tests to use 4 procs

Refs #26405

lindsayad commented:
requires a libmesh update

lindsayad committed:
Properly set reinit_displaced members

lindsayad committed:
Move SwapBackSentinel to proper place in NonlinearThread

lindsayad committed:
Setup swap sentinels in the HybridizedKernel

lindsayad committed:
Default to storing the passed-in predicate by value

But have the option to store it as a reference in the case that a
user may want to change the boolean in the swap back sentinel's
scope of duration.

Pass by value must be the default because previous user code
passed in temporaries

lindsayad committed:
Give hybridized objects their own tasks and bases

This prevents regular Kernels from being placed under the
`HybridizedKernels` syntax which would probably lead to bad things

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2039049/) on c82a664 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/26477/site)

This comment will be updated on new commits.

lindsayad committed:
Add deprecated BoundaryID overload for reinitElemFace

moosebuild commented:
Job [Modules parallel](https://civet.inl.gov/job/2039391/) on fc7dc8d : invalidated by @lindsayad

lindsayad committed:
Remove testing redundancy in JSON TestSearch

Each invocation of an oprof MOOSE executable with `--json` takes
about 5 seconds. It's far far worse when compiled with `-O0`. Both
of the queries in the `TestSearch` test were looking at "blocks" type
data. So there is no need to make this test run 2X as long to test
the same thing 2X

lindsayad committed:
Add documentation

lindsayad committed:
Change Mixed to Primal

I think this is important to do as Rhebergen introduces hybridized
methods that don't introduce additional variables for gradients

lindsayad committed:
Remove _lm_size/_primal_size from framework hybridized classes

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2058940/) on 0a4caef wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>8f32c4</th>
      <th colspan=4>#26477 0a4cae</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.26%</td>
      <td>85.12%</td>
      <td>-0.14%</td>
      <td>40.22%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>100412</td>
      <td>100440</td>
      <td>+28</td>
      <td>144</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>17359</td>
      <td>17562</td>
      <td>+203</td>
      <td>214</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/26477/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/26477/coverage/framework)


## Modules coverage

### Navier stokes

<table>
  <thead>
    <tr>
      <th></th>
      <th>8f32c4</th>
      <th colspan=4>#26477 0a4cae</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.76%</td>
      <td>85.13%</td>
      <td>+0.37%</td>
      <td>98.34%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>14577</td>
      <td>15052</td>
      <td>+475</td>
      <td>475</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>2621</td>
      <td>2629</td>
      <td>+8</td>
      <td>8</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/26477/coverage_diff/navier_stokes)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/26477/coverage/navier_stokes)

### Phase field

<table>
  <thead>
    <tr>
      <th></th>
      <th>8f32c4</th>
      <th colspan=4>#26477 0a4cae</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.64%</td>
      <td>85.64%</td>
      <td>-</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>13132</td>
      <td>13132</td>
      <td>-</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>2202</td>
      <td>2202</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/26477/coverage_diff/phase_field)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/26477/coverage/phase_field)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/26477/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/26477/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/26477/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/26477/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/26477/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/26477/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/26477/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/26477/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/26477/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/26477/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/26477/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/26477/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/26477/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/26477/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/26477/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/26477/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/26477/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/26477/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/26477/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/26477/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/26477/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/26477/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/26477/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/26477/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/26477/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/26477/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/26477/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/26477/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 40.22% is less than the suggested 90.0%



This comment will be updated on new commits.

lindsayad committed:
Change BC virtual method from assemble to onBoundary

This is consistent with our loop method names and felt better when
writing the documentation

lindsayad committed:
Move hybridized objects into their own directories

lindsayad committed:
Create framework level hybridized diffusion objects and test

lindsayad committed:
Migrate the weak forms from DiffusionHybridizedKernel to its interface class

lindsayad committed:
Move diffusion Dirichlet routines into interface class

This is with the plan to reuse these routines in NS

lindsayad committed:
Reuse code from DiffusionHybridizedInterface in NS hybridized classes

lindsayad committed:
Migrate weak forms from NS hybridized kernel/bcs into interface

This makes it consistent with the diffusion code: all the physics
are defined in the interface classes

lindsayad committed:
Add MMS test for hybridized diffusion

lindsayad committed:
Use boomeramg for hybridized diffusion tests

So the matrix is nonsymmetric which I need to investigate since the
corresponding libMesh example does have a symmetric matrix. My guess
is to do with the single face hybridizable, which is what the libMesh
example is doing, but the MOOSE example is not. However, the MOOSE
test still appears to be SPD "enough" for boomeramg to work well

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2063215/) on d4d3c6d wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/26477/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/26477/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format c40f5a62156469f2a42a8dc81e4cf8f1c67bbfa3`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2063267/) on 6c1df77 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/26477/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/26477/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format c40f5a62156469f2a42a8dc81e4cf8f1c67bbfa3`

lindsayad committed:
Documentation for framework objects

GiudGiud commented:
Can we narrow down the name? Maybe HybridDG?

lindsayad commented:
I suppose

lindsayad commented:
Actually I think we are narrowing the name down I would do with HDG. Hybrid can have a lot of meanings. In the literature what I am implementing is almost always referred to has hybridized or hybridizable. HDG would be very recognizable to someone familiar with the field

GiudGiud commented:
Works for me

lindsayad committed:
Hybridized -> HDG

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2067671/) on 53f9841 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/26477/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/26477/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 7cc4aedd3b71997307aaaf7cbae1810fd32698eb`

GiudGiud commented on framework/include/interfaces/DiffusionHDGInterface.h:
```diff
@@ -0,0 +1,476 @@
+  const std::vector<dof_id_type> & _u_dof_indices;
+  const std::vector<dof_id_type> & _lm_u_dof_indices;
+
+  // local solutions at quadrature points
```
maybe not. Maybe we should just keep everything local for simplicity

GiudGiud commented on framework/include/interfaces/DiffusionHDGInterface.h:
```diff
@@ -0,0 +1,476 @@
+      // Vector equation dependence on scalar dofs
+      for (const auto j : make_range(obj._scalar_n_dofs))
+        obj._PrimalMat(i_offset + i, scalar_j_offset + j) +=
+            obj._JxW[qp] * (obj._div_vector_phi[i][qp] * obj._scalar_phi[j][qp]);
```
is there an equation in the doc we can tie to this?

lindsayad replied:
yes, adding some

GiudGiud commented on framework/src/actions/AddAuxVariableAction.C:
```diff
@@ -28,7 +28,10 @@ AddAuxVariableAction::AddAuxVariableAction(const InputParameters & params)
 AddAuxVariableAction::getAuxVariableFamilies()
 {
-  return MooseEnum("LAGRANGE MONOMIAL SCALAR LAGRANGE_VEC MONOMIAL_VEC", "LAGRANGE", true);
+  return MooseEnum("LAGRANGE MONOMIAL SCALAR LAGRANGE_VEC MONOMIAL_VEC L2_HIERARCHIC "
```
missing L2_LAGRANGE?
or maybe I cant see it, why is it being accepted later

GiudGiud commented on framework/src/hdgbcs/DiffusionHDGDirichletBC.C:
```diff
@@ -0,0 +1,49 @@
+  auto params = HDGIntegratedBC::validParams();
+  params += DiffusionHDGInterface::validParams();
+  params.addClassDescription("Weakly imposes Dirichlet boundary conditions for a "
+                             "hybridized discretization of diffusion equation");
```
```suggestion
                             "hybridized discretization of a diffusion equation");
```

GiudGiud commented on framework/src/hdgbcs/DiffusionHDGDirichletBC.C:
```diff
@@ -0,0 +1,49 @@
+  params += DiffusionHDGInterface::validParams();
+  params.addClassDescription("Weakly imposes Dirichlet boundary conditions for a "
+                             "hybridized discretization of diffusion equation");
+  params.addParam<FunctionName>("function", 0, "The Dirichlet value for the diffusing specie");
```
any chance this could be a functor?
would help connect HDG flow to other flows in the future

GiudGiud commented on framework/src/hdgbcs/DiffusionHDGZeroFluxBC.C:
```diff
@@ -0,0 +1,45 @@
+  auto params = HDGIntegratedBC::validParams();
+  params += DiffusionHDGInterface::validParams();
+  params.addClassDescription("Implements a zero flux boundary condition for use with a hybridized "
+                             "discretization of the diffusion equation");
```
what s the hold up on non-zero flux?

GiudGiud commented on framework/include/loops/PreCheckThread.h:
```diff
@@ -0,0 +1,42 @@
+  void join(const PreCheckThread &) {}
+
+protected:
+  /// the hybridized kernels we will call the pre-check method for
```
it s not called pre-check anymore? post-something now?

lindsayad replied:
I don't understand the question. There are both pre-line-search and post-line-search callbacks from PETSc (precheck and postcheck)

GiudGiud commented on modules/navier_stokes/doc/content/source/hdgbcs/NavierStokesHDGOutflowBC.md:
```diff
@@ -0,0 +1,17 @@
+# NavierStokesHDGOutflowBC
```
add HDG to index of all discretizations on main page

GiudGiud commented on modules/navier_stokes/doc/content/source/hdgbcs/NavierStokesHDGOutflowBC.md:
```diff
@@ -0,0 +1,17 @@
+\end{equation}
+
+where $\vec{n}$ is the outward-facing normal on the boundary, $p$ is the
+pressure, $\nu$ is the kinematic viscosity, and $\vec{u}$ is the velocity.
```
ugh all the fluid properties use mu (dynamic). can you use mu?

grmnptr replied:
Are you sure it is the kinematic viscosity? In some incompressible simulations people divide by the constant rho and then this is true but then p is not the real pressure.

GiudGiud replied:
@lindsayad

GiudGiud replied:
maybe we should just have KinematicViscosityFunctorMaterial ? and we ll use that for HDG?

lindsayad replied:
The current form is divided through by rho. I will change it tho

lindsayad replied:
changed

GiudGiud commented on framework/src/systems/NonlinearSystemBase.C:
```diff
@@ -500,6 +502,38 @@ NonlinearSystemBase::addKernel(const std::string & kernel_name,
+                                  InputParameters & parameters)
+{
+  // The hybridized objects require that the residual and Jacobian be computed together
+  residualAndJacobianTogether();
```
need that too for adding the BC?

GiudGiud commented on modules/navier_stokes/include/interfaces/NavierStokesHDGInterface.h:
```diff
@@ -0,0 +1,698 @@
+
+#include "DiffusionHDGInterface.h"
+
+class NavierStokesHDGInterface : public DiffusionHDGInterface
```
i see that s why you used nu

GiudGiud reviewed (commented):
did a first pass while procrastinating other items

GiudGiud commented on framework/include/interfaces/HDGInterface.h:
```diff
@@ -0,0 +1,32 @@
+typedef Eigen::VectorXd EigenVector;
+#endif
+
+class HDGInterface
```
is interface the right name / system for these classes?
what is it interfacing between? The system and the ROs?

lindsayad replied:
It's a common interface for both hybridized kernels and boundary conditions

GiudGiud replied:
Sure but not every common class has to be called an interface

GiudGiud replied:
If you mean interface to the Assembly then it's fine. But it could be also `HDGAssemblyBase` or `HDGAssemblyInterface` or `HDGAssemblyHelper`. 

`Base` somewhat implies a lot of the other interfaces will be included
anyway this is a side topic, dont view this as a blocking conversation

lindsayad replied:
I kind of like XXXAssemblyInterface

lindsayad replied:
In this case I'm going to call it `HDGData` since it just supplies data members and no method interfaces. But for the "interfaces" with meat to them I'm going to go with XXXAssemblyInterface

lindsayad replied:
Err, I think I like XXXHDGAssemblyHelper. I'll probably switch again, we'll have to see haha

grmnptr commented on framework/include/systems/NonlinearSystemBase.h:
```diff
@@ -145,6 +147,26 @@ class NonlinearSystemBase : public SystemBase, public PerfGraphInterface
                          InputParameters & parameters);
 
+  /**
+   * Adds a hybridized kernel
```
```suggestion
   * Adds a Hybridized Discontinuous Galerkin kernel
```

grmnptr replied:
feel free to disregard this and the next few comments if hybridized is already implies DG

grmnptr commented on framework/include/systems/NonlinearSystemBase.h:
```diff
@@ -145,6 +147,26 @@ class NonlinearSystemBase : public SystemBase, public PerfGraphInterface
+                            InputParameters & parameters);
+
+  /**
+   * Adds a hybridized bc
```
```suggestion
   * Adds a Hybridized Discontinuous Galerkin boundary condition
```

grmnptr commented on framework/include/systems/NonlinearSystemBase.h:
```diff
@@ -819,6 +841,7 @@ class NonlinearSystemBase : public SystemBase, public PerfGraphInterface
   ///@{
   /// Kernel Storage
   MooseObjectTagWarehouse<KernelBase> _kernels;
+  MooseObjectWarehouse<HDGKernel> _hybridized_kernels;
```
This is more of a question based on what I see for finite volume objects. Which one is preferred:
1. Using object warehouses forevery types
2. Putting them into TheWarehouse
I started using the thewarehouse on a different PR, but if (1) is preferred I'll just transition to using object warehouses. And maybe we could do the same for FV too.

grmnptr commented on framework/src/hdgbcs/DiffusionHDGDirichletBC.C:
```diff
@@ -0,0 +1,49 @@
+{
+  resizeData(*this);
+
+  // qu, u
```
Linking to the paper with a comment might be important here to know what qu and u are. Same for others below.

grmnptr commented on framework/src/loops/PreCheckThread.C:
```diff
@@ -0,0 +1,67 @@
+{
+  _fe_problem.subdomainSetup(_subdomain, _tid);
+
+  // Update variable Dependencies
```
```suggestion
  // Update variable dependencies
```

grmnptr reviewed (commented):
Interesting work, posting through the framework parts.

lindsayad commented on test/tests/outputs/format/test_json.py:
```diff
@@ -170,15 +170,6 @@ def testJsonSearch(self):
-        self.assertIn("BadKernels", data)
-        self.assertIn("Kernels", data)
-        diff = data["Kernels"]["star"]["subblock_types"]["Diffusion"]
-        self.assertIn("use_displaced_mesh", diff["parameters"])
```
see https://github.com/idaholab/moose/pull/26477/commits/0a4caefd4877fd46370a5afbacc77053071e1c57

lindsayad committed:
Address Guillaume review

- Remove unecessary forward declarations
- Doxygen
- Typo fixes
- More const in header method declarations
- Documentation for the HDG diffusion weak forms
- Add L2_LAGRANGE to available aux families

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

lindsayad commented on framework/doc/content/syntax/HDGKernels/index.md:
```diff
@@ -0,0 +1,66 @@
+[HDGBCs/index.md].
+
+Within `onElement` and `onInternalSide`, hybridized kernel developers have eight
+different data structures they need to populate. These are
```
they are protected data members of the class

lindsayad commented on framework/include/hdgkernels/HDGKernel.h:
```diff
@@ -0,0 +1,179 @@
+
+  virtual void initialSetup() override;
+
+  static const std::string lm_increment_vector_name;
```
No; there are three different Lagrange multiplier variables in a 2D NS HDG sim. Pressure, LM u, LM v. And some HDG discretizations add an LM p as well. This is the name of the `NumericVector` that holds the changes in all these LMs from one Newton iteration to the next

lindsayad commented on framework/include/loops/PreCheckThread.h:
```diff
@@ -0,0 +1,42 @@
+class MooseObjectWarehouse;
+class HDGKernel;
+
+class PreCheckThread : public ThreadedElementLoop<ConstElemRange>
```
at some point someone might want to put something in here that is unrelated to HDG, but I suppose we could cross that bridge when we get there

lindsayad replied:
Bingo

lindsayad replied:
doxygen added and renamed

lindsayad committed:
HDGInterface -> HDGData

lindsayad committed:
XXXHDGInterface -> XXXHDGAssemblyHelper

lindsayad committed:
Use functor for diffusion Dirichlet value

lindsayad committed:
tests/hybridizedkernels -> tests/hdgkernels

lindsayad committed:
DiffusionHDGZeroFluxBC -> DiffusionHDGPrescribedGradientBC

lindsayad committed:
Add HDG to main Navier-Stokes page

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2114416/) on 4fe90d9 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/26477/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/26477/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 7e9237252a8529c78c1df88067012d16210dfe6a`

lindsayad committed:
Add a nonzero prescribed gradient test

This necessitated removing the `if (obj._neigh)` guard in the
scalar face residual/Jacobian computation. Perhaps unsurprisingly
this appears to have fixed my asymmetric matrix issue. Indeed if
I am going to have the LM equation active on a prescribed gradient/flux
boundary then the transpose terms should also be present (e.g. we
should continue to add the corresponding terms in the scalar equation)

lindsayad committed:
Address Peter review

- Clarifying note on what the mesh skeleton is
- Clearer doxygen for system HDG add methods
- Add Cockburn's article for notation reference in code comments
- Other word smithing

lindsayad commented on framework/include/hdgbcs/HDGIntegratedBC.h:
```diff
@@ -0,0 +1,118 @@
+  const Elem * _neigh = nullptr;
+
+private:
+  friend class HDGKernel;
```
The HDGKernel is the central repository for elemental contributions. It will actually loop over the HDG BCs and its residual/Jacobian contributions to the central repository. So it accesses HDGBC data. See:
```c++
void
HDGKernel::addBCData(const HDGIntegratedBC & hibc)
{
  _PrimalMat += hibc._PrimalMat;
  _LMMat += hibc._LMMat;
  _PrimalLM += hibc._PrimalLM;
  _LMPrimal += hibc._LMPrimal;
  _PrimalVec += hibc._PrimalVec;
  _LMVec += hibc._LMVec;
}
```
The reason we do this here and not in, say, the `Assembly` class, is that only the physics knows what fields are primal vs. LM fields and hence how to condense the primal data into LM data before adding to the global residual/Jacobian data structures which contain only the LM degrees of freedom

lindsayad commented on framework/include/interfaces/DiffusionHDGInterface.h:
```diff
@@ -0,0 +1,476 @@
+void
+DiffusionHDGInterface::resizeData(DiffusionHDG & obj)
+{
+  obj._vector_n_dofs = obj._qu_dof_indices.size();
```
It does allow for this base "interface" to access protected/private members of derived classes

lindsayad replied:
because of the diamond problem I cannot have both `NavierStokesHDGAssemblyHelper` inherit from `DiffusionHDGAssemblyHelper` and `NavierStokesHDGKernel` inherit from `DiffusionHDGKernel`. I'm pretty sure that's the core reason of why things end up looking a bit weird

lindsayad commented on framework/include/interfaces/DiffusionHDGInterface.h:
```diff
@@ -0,0 +1,476 @@
+
+template <typename DiffusionHDG>
+void
+DiffusionHDGInterface::scalarVolumeResidual(DiffusionHDG & obj,
```
I'm hopeful that the doxygen I added per @GiudGiud's earlier comment helps in this regard

lindsayad committed:
Add tensor-product element MMS test

lindsayad committed:
Fix NS outflow

We restored symmetry for the diffusion hybridization at flux boundaries
so I must also do the same for the NS hybridization at flux boundaries

grmnptr commented on modules/navier_stokes/doc/content/source/hdgkernels/NavierStokesHDGKernel.md:
```diff
@@ -0,0 +1,21 @@
+\nabla \cdot \left(\vec{u} \otimes \vec{u}\right) - \nabla \cdot \left(\nu
+\nabla\vec{u}\right) + \nabla p = 0
+
+where $\nu$ is the kinematic viscosity and $p$ is the pressure. This class uses
```
```suggestion
where $\nu$ is the kinematic viscosity and $p$ is the density-corrected pressure. This class uses
```

grmnptr replied:
Assuming that you want to keep this formulation.

lindsayad replied:
changed to use rho and mu

grmnptr commented on modules/navier_stokes/include/hdgbcs/NavierStokesHDGOutflowBC.h:
```diff
@@ -0,0 +1,42 @@
+
+/**
+ * Implements an outflow boundary condition for use with a hybridized discretization of the
+ * Navier-Stokes equations
```
```suggestion
 * incompressible Navier-Stokes equations
```

grmnptr replied:
If you keep the mu/rho formulation then this is only valid for incompressible.

grmnptr commented on modules/navier_stokes/include/hdgkernels/NavierStokesHDGAssemblyHelper.h:
```diff
@@ -0,0 +1,698 @@
+
+#include "DiffusionHDGAssemblyHelper.h"
+
+class NavierStokesHDGAssemblyHelper : public DiffusionHDGAssemblyHelper
```
Docstring to describe what this class is meant to do.

grmnptr commented on modules/navier_stokes/src/hdgbcs/NavierStokesHDGOutflowBC.C:
```diff
@@ -0,0 +1,73 @@
+NavierStokesHDGOutflowBC::NavierStokesHDGOutflowBC(const InputParameters & parameters)
+  : HDGIntegratedBC(parameters),
+    NavierStokesHDGAssemblyHelper(this, this, _sys, _aux_sys, _mesh, _tid)
+{
```
Check if the corresponding NS kernel is there.

grmnptr commented on modules/navier_stokes/src/hdgbcs/NavierStokesHDGOutflowBC.C:
```diff
@@ -0,0 +1,73 @@
+  auto params = HDGIntegratedBC::validParams();
+  params += NavierStokesHDGAssemblyHelper::validParams();
+  params.addClassDescription("Implements an outflow boundary condition for use with a hybridized "
+                             "discretization of the Navier-Stokes equations");
```
```suggestion
                             "discretization of the incompressible Navier-Stokes equations");
```

grmnptr commented on modules/navier_stokes/src/hdgbcs/NavierStokesHDGVelocityDirichletBC.C:
```diff
@@ -0,0 +1,74 @@
+    const InputParameters & parameters)
+  : HDGIntegratedBC(parameters),
+    NavierStokesHDGAssemblyHelper(this, this, _sys, _aux_sys, _mesh, _tid)
+{
```
Check if the corresponding kernel is there.

grmnptr commented on modules/navier_stokes/test/tests/finite_element/ins/hdg/lid.i:
```diff
@@ -0,0 +1,88 @@
+  []
+[]
+
+[HDGKernels]
+  [ns]
+    type = NavierStokesHDGKernel
+  []
+[]
+
+[HDGBCs]
+  [walls]
+    type = NavierStokesHDGVelocityDirichletBC
+    boundary = 'left right bottom'
+  []
+  [lid]
+    type = NavierStokesHDGVelocityDirichletBC
+    boundary = 'top'
+    dirichlet_u = '1'
+  []
+[]
```
Very nice and concise syntax!

grmnptr reviewed (changes requested):
Nice! Just a few minor comments/questions. This is my last review.

GiudGiud commented on framework/src/hdgkernels/HDGKernel.C:
```diff
@@ -0,0 +1,176 @@
+      onInternalSide();
+    }
+  }
+
```
there's two parts to this assemble routine

let s split them with a comment like:
// Now that the HDGData members are filled, we contribute them to ..

GiudGiud commented on framework/src/hdgkernels/HDGKernel.C:
```diff
@@ -0,0 +1,176 @@
+
+    // We were performing our finite element assembly for the implicit solve step of our
+    // example. Add our local element vectors/matrices into the global system
+    addResiduals(_assembly, _F_libmesh, _lm_dof_indices, /*scaling_factor=*/1);
```
why one for scaling factors? this is my ignorance again but shoujld we use the variable scaling factor

lindsayad replied:
scaling factors would disrupt the relationship between the LM and primal solution increments. I will add an error message

GiudGiud commented on framework/src/hdgkernels/HDGKernel.C:
```diff
@@ -0,0 +1,176 @@
+    // We are doing our finite element assembly for the second time. We now know the Lagrange
+    // multiplier solution. With that and the local element matrices and vectors we can compute
+    // the vector and scalar solutions
+    //
```
so which loop are in now? ComputeElemAux ?
and which loop are we in the first statement of this if condition ? the PreCheck ?

lindsayad replied:
I will add a comment. It's the `ComputeResidualAndJacobianThread` for the former and the now `HDGPrimalSolutionUpdateThread` for the latter

GiudGiud commented on framework/src/actions/AddAuxVariableAction.C:
```diff
@@ -28,7 +28,10 @@ AddAuxVariableAction::AddAuxVariableAction(const InputParameters & params)
 {
-  return MooseEnum("LAGRANGE MONOMIAL SCALAR LAGRANGE_VEC MONOMIAL_VEC", "LAGRANGE", true);
+  return MooseEnum("LAGRANGE MONOMIAL SCALAR LAGRANGE_VEC MONOMIAL_VEC L2_HIERARCHIC "
+                   "L2_HIERARCHIC_VEC L2_LAGRANGE L2_LAGRANGE_VEC",
```
so does auxkernels work with these variables? I remember some trouble while working on projectionAux

lindsayad replied:
they should be able to. That doesn't mean I'm going to guarantee that all aux kernels work with them

GiudGiud commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -2517,10 +2525,37 @@ FEProblemBase::addKernel(const std::string & kernel_name,
+      determineNonlinearSystem(parameters.varName("variable", name), true).second;
+  setResidualObjectParamsAndLog(
+      kernel_name, name, parameters, nl_sys_num, "Kernel", _reinit_displaced_elem);
+
```
add the logAdd call for debugging

lindsayad replied:
that's done in `setResidualObjectParamsAndLog`

GiudGiud commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -2517,10 +2525,37 @@ FEProblemBase::addKernel(const std::string & kernel_name,
+      determineNonlinearSystem(parameters.varName("variable", name), true).second;
+  setResidualObjectParamsAndLog(
+      kernel_name, name, parameters, nl_sys_num, "HDGKernel", _reinit_displaced_elem);
+
```
same

lindsayad replied:
that's done in `setResidualObjectParamsAndLog`

GiudGiud commented on framework/doc/content/syntax/HDGKernels/index.md:
```diff
@@ -0,0 +1,66 @@
+hybridization as primal variables and the variable(s) that live on the mesh
+skeleton as Lagrange multipliers (LMs) or dual variable(s).
+
+## Implementation in MOOSE
```
where would be a good place to say there are 'ran' twice per iteration?

GiudGiud commented on modules/navier_stokes/include/hdgkernels/NavierStokesHDGAssemblyHelper.h:
```diff
@@ -0,0 +1,698 @@
+      // Scalar equation dependence on scalar dofs
+      for (const auto j : make_range(obj._scalar_n_dofs))
+      {
+        // derivatives wrt 0th component
```
0th component of ?

GiudGiud reviewed (changes requested):
looks good. I felt a little better going through the code this time, thanks for adding these comments

lindsayad committed:
Add a 'IsMatrixSymmetric' postprocessor

lindsayad committed:
Make IsMatrixSymmetric more robust

lindsayad commented:
Saving a picture here for Re = 10k

<img width="573" alt="Screenshot 2024-04-01 at 9 46 25 PM" src="https://github.com/idaholab/moose/assets/10248304/a4548c22-9767-4817-99b5-9d1a22a434bb">

lindsayad committed:
Update addHDG APIs to use determineSolverSystem

lindsayad committed:
PreCheckThread -> HDGPrimalSolutionUpdateThread

lindsayad committed:
Change from using nu to rho and mu

lindsayad committed:
Add framework symmetric matrix test for HDG Diffusion

lindsayad committed:
Add symmetric stokes testing

lindsayad committed:
Avoid templating/friend design for HDG framework objects

lindsayad committed:
Avoid templating/friend design for HDG NS objects

lindsayad commented on framework/include/hdgbcs/DiffusionHDGZeroFluxBC.h:
```diff
@@ -0,0 +1,38 @@
+protected:
+  virtual void onBoundary() override;
+
+  friend class DiffusionHDGInterface;
```
template/friend design has been removed

lindsayad committed:
Address Peter review

- Improved doxygen
- Error if kernel/bcs operating on different variables
- Error if kernels and boundary conditions aren't matching physics

lindsayad commented on framework/include/hdgkernels/HDGKernel.h:
```diff
@@ -0,0 +1,179 @@
+  /// The physical locations of the quadrature points on the face
+  const MooseArray<Point> & _q_point_face;
+
+  /// transformed Jacobian weights on the face
```
I don't think there's an easy way to. The only way I can think is to add an API that takes a quadrature point index and returns `JxW[qp]` that a user has to call for every qp

lindsayad commented on framework/include/hdgkernels/HDGKernel.h:
```diff
@@ -0,0 +1,179 @@
+  /// The warehouse holding the hybridized integrated boundary conditions
+  MooseObjectWarehouse<HDGIntegratedBC> & _hibc_warehouse;
+
+  /// Lagrange multiplier matrix and RHS after eliminating vector and scalar dofs
+  DenseMatrix<Number> _K_libmesh;
+  DenseVector<Number> _F_libmesh;
+
+  /// Primal matrix inverse
+  EigenMatrix _PrimalMatInv;
+
+  // local degree of freedom increment values
+  std::vector<Number> _lm_increment_dof_values;
+  std::vector<Number> _primal_increment_dof_values;
+  EigenVector _LMIncrement, _PrimalIncrement;
```
> HDGData limited to stuff shared between BCs and kernels
correct

lindsayad committed:
Address Guillaume review

- more doxygen
- more code comments
- use functors for source in diffusion problems

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2223464/) on 1bce4f7 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/26477/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/26477/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 639315765d5bfb1a99ac58caacdb50a4a3d7ee22`

lindsayad commented on framework/doc/content/source/hdgkernels/DiffusionHDGKernel.md:
```diff
@@ -0,0 +1,26 @@
+weak forms implemented are a slight variation on the cited work in order to make
+them directly usable in downstream physics such as Navier-Stokes.
+
+\begin{equation}
```
yes, it's the third one

lindsayad committed:
Markdown page for IsMatrixSymmetric

lindsayad commented:
Ok this thing should be ready for review again

moosebuild commented:
Job [Modules parallel](https://civet.inl.gov/job/2225104/) on 8dbd4d7 : invalidated by @lindsayad

GiudGiud commented on framework/include/hdgkernels/HDGData.h:
```diff
@@ -0,0 +1,42 @@
+typedef Eigen::VectorXd EigenVector;
+#endif
+
+class HDGData
```
```suggestion
/**
  * Class to hold data structures shared between kernels and boundary conditions
  */
class HDGData
```

GiudGiud commented on framework/include/hdgkernels/HDGKernel.h:
```diff
@@ -0,0 +1,175 @@
+inline void
+HDGKernel::computeResidual()
+{
+  // Would be nice to error for this but at the time of writing we still do a usually unnecessary
+  // initial residual evaluation
```
times have changed...

GiudGiud commented on framework/include/postprocessors/IsMatrixSymmetric.h:
```diff
@@ -0,0 +1,38 @@
+class PetscMatrix;
+}
+
+class IsMatrixSymmetric : public GeneralPostprocessor
```
```suggestion
/**
  * Checks whether the nonlinear system matrix is symmetric
  */
class IsMatrixSymmetric : public GeneralPostprocessor
```

GiudGiud commented on framework/src/postprocessors/IsMatrixSymmetric.C:
```diff
@@ -0,0 +1,77 @@
+{
+  InputParameters params = GeneralPostprocessor::validParams();
+  params.addClassDescription("Report whether the system matrix is symmetric or not.");
+  params.addParam<Real>("symmetry_tol", 1e-8, "The tolerance for comparing symmetry");
```
```suggestion
  params.addParam<Real>("symmetry_tol", 1e-8, "The tolerance (both relative and absolute) for comparing symmetry");
```

GiudGiud commented on framework/include/postprocessors/IsMatrixSymmetric.h:
```diff
@@ -0,0 +1,38 @@
+  virtual Real getValue() const override;
+
+protected:
+  const Real _symm_tol;
+  std::unique_ptr<SparseMatrix<Number>> _mat_transpose;
+  bool _equiv = true;
```
```suggestion
  /// Tolerance for the comparison between coefficients and their symmetric
  const Real _symm_tol;
  /// Transpose of the system matrix
  std::unique_ptr<SparseMatrix<Number>> _mat_transpose;
  /// Whether the matrix is symmetric
  bool _equiv = true;
```

GiudGiud commented on framework/src/postprocessors/IsMatrixSymmetric.C:
```diff
@@ -0,0 +1,77 @@
+    {
+      const auto val1 = sys_mat(i, j);
+      const auto val2 = (*_mat_transpose)(i, j);
+      if (!MooseUtils::relativeFuzzyEqual(val1, val2, _symm_tol) &&
```
```suggestion
      if (!MooseUtils::relativeFuzzyEqual(val1, val2, _symm_tol) ||
```
more conservative

lindsayad replied:
I deliberately made this an or because this failed for an entry on the order of machine precision

GiudGiud replied:
alright

GiudGiud reviewed (commented):
getting back to it in a few

GiudGiud commented on framework/src/hdgkernels/HDGKernel.C:
```diff
@@ -0,0 +1,235 @@
+    _hibc_warehouse(
+        *getCheckedPointerParam<MooseObjectWarehouse<HDGIntegratedBC> *>("hibc_warehouse"))
+{
+  // This class handles residuals and Jacobians for multiple variables
```
HDG diffusion doesn't
might be worth moving this to NSHDG

GiudGiud commented on modules/navier_stokes/doc/content/source/hdgbcs/NavierStokesHDGOutflowBC.md:
```diff
@@ -0,0 +1,17 @@
+hybridized kernel [NavierStokesHDGKernel.md]. The condition imposed is
+
+\begin{equation}
+\vec{n}p -\nu\nabla\vec{u}\cdot{n} = \vec{0}
```
```suggestion
\vec{n}p -\nu\nabla\vec{u}\cdot{n} = \vec{0}
```
didnt you switch to mu?
also diffusion term should be a vector too

GiudGiud reviewed (commented):
just the two assembly helper left. I ll take another look on a more rested day.

you mention raviart thomas in the doco but all the examples (iirc) are hierarchic or lagrange. any isssue with RT?

grmnptr commented on framework/doc/content/syntax/HDGKernels/index.md:
````diff
@@ -0,0 +1,77 @@
+different data structures they need to populate. These are
+
+```
+  /// Matrix data structures for on-diagonal coupling
+  EigenMatrix _PrimalMat, _LMMat;
+  /// Vector data structures
+  EigenVector _PrimalVec, _LMVec;
+  /// Matrix data structures for off-diagonal coupling
+  EigenMatrix _PrimalLM, _LMPrimal;
````
Maybe you could mention that some of these are in the `HDGData` class now. I like the design!

grmnptr commented on modules/navier_stokes/test/tests/finite_element/ins/hdg/lid.i:
```diff
@@ -0,0 +1,89 @@
+[Executioner]
+  type = Steady
+  solve_type = NEWTON
+  petsc_options_iname = '-pc_type -pc_factor_shift_type'
+  petsc_options_value = 'lu       NONZERO'
```
One of the reasonswhy HDG might be so good is that you can do field split preconditioning with it (if I remember correctly). I think it would be beneficial to the users to see an example with that. Unless there is already one, in which case feel free to resolve.

lindsayad replied:
HDG is no more amenable to field split than other finite element discretizations. It's draw is that it gives the same benefits of a DG (or FV) discretization in terms of local element conservation and natural ability to upwind convection schemes while allowing you to condense out the internal degrees of freedom that make a DG scheme have substantially more global degrees of freedom than a Continuous Galerkin discretization.

In at least one way HDG is *harder* to field split than CG. It takes a larger gamma coefficient in grad-div stabilization in order to make the Schur complement spectrally equivalent to the pressure mass matrix. This larger perturbation makes the momentum equation have worse conditioning. This is what I am currently researching. You can see https://gitlab.com/petsc/petsc/-/merge_requests/7550 for some work in that arena. 

All that is to say that field split is still a fair amount of work. It actually requires *not* condensing out the internal dofs, at least not until you are working on the block velocity level. I do have a field split example in https://github.com/idaholab/moose/pull/27188.

grmnptr replied:
Great, sounds good to me!

grmnptr reviewed (commented):
I have two more comments but it looks good to me. My approval is given whenever Guillaume's comments are addressed. Great contribution!

lindsayad commented on framework/doc/content/source/hdgbcs/DiffusionHDGPrescribedGradientBC.md:
```diff
@@ -0,0 +1,19 @@
+prescribes the normal gradient
+
+\begin{equation}
+\vec{n} \cdot D \nabla u = D q_N
```
why bold?

GiudGiud replied:
it s vector once you have taken the gradient

lindsayad replied:
In the overwhelming majority of occurrences in our documentation we do not bold face the del operator regardless of whether its result is a gradient, divergence, etc.

grmnptr replied:
Yeah I think bfing nabla is not needed, it is the derivative vector by definition.

lindsayad committed:
Address Guillaume and Peter reviews

- More doxygen
- Documentation fixes/additions
- Error for computeResidual call
- Move kernel coverage call to NavierStokesHDGKernel

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

lindsayad commented on framework/include/hdgkernels/DiffusionHDGAssemblyHelper.h:
```diff
@@ -0,0 +1,262 @@
+  std::size_t _primal_size, _lm_size;
+
+  /// Our stabilization coefficient
+  static constexpr Real _tau = 1;
```
Yes. See in the `scalarFaceResidual` method for example

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2233407/) on daa430b wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/26477/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/26477/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format f1a3d3323ab5a0b522f3fb2a55bd7e9bc3b51547`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2233461/) on a418f06 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/26477/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/26477/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format f1a3d3323ab5a0b522f3fb2a55bd7e9bc3b51547`

lindsayad committed:
Add Raviart-Thomas test

lindsayad commented:
Hopefully read again

GiudGiud commented on framework/doc/content/source/hdgkernels/DiffusionHDGKernel.md:
```diff
@@ -0,0 +1,29 @@
+Neumann boundary, $\partial\Omega_N$. As indicated by the test functions, the
+first equation is the equation for the variable $\vec{q}$, the second is for the
+variable $u$, and the third is for the trace or Lagrange multiplier variable
+$\hat{u}$.
```
```suggestion
$\hat{u}$.

$\vec{q}$ (the gradient) and $u$ (the values) are referred to as the primal variables. They can be summed to construct the value of the diffused quantity over the domain. $\hat{u} is often referred to as the dual variable. Its equation is the only one that requires the use of a linear solver. For more information on HDG solves, see the [HDGKernels page](HDGKernels/index.md).  
```

GiudGiud replied:
summed is probably not the right word

GiudGiud replied:
can $q$ be computer from $u$ as its gradient? Or do you have to solve that equation to compute it?

lindsayad replied:
In this class of HDG discretizations, the gradient is an independent variable that must be solved for. You cannot compute it from u. There are other HDG methods (like the interior penalty one in the other PR) that do not solve for the gradient. Solution of the gradient does allow you to postprocess even higher order scalar solution approximations (e.g. k + 2 instead of the normal k + 1 for L2 norm convergence) for diffusion or diffusion-dominated problems. Once you become advection dominated though, the postprocessing is no longer effective so you lose that benefit

lindsayad replied:
```suggestion
$\hat{u}$.

$\vec{q}$ (the gradient) and $u$ (the values) are referred to as the primal variables. $\hat{u} is often referred to as the dual variable. Its equation is the only one that requires the use of a linear solver. For more information on HDG solves, see the [HDGKernels page](HDGKernels/index.md).  
```

GiudGiud commented on framework/doc/content/source/hdgkernels/DiffusionHDGKernel.md:
```diff
@@ -0,0 +1,29 @@
+first equation is the equation for the variable $\vec{q}$, the second is for the
+variable $u$, and the third is for the trace or Lagrange multiplier variable
+$\hat{u}$.
+
```
```suggestion

$f$ the forcing function can be used for method of manufactured solutions studies, or as a source term for the diffused quantity.

```

GiudGiud commented on framework/include/hdgkernels/DiffusionHDGAssemblyHelper.h:
```diff
@@ -0,0 +1,262 @@
+
+  /**
+   * Computes a local residual vector for the weak form:
+   * -<Dq*n, w> + <\tau * (u - \hat{u}) * n * n, w>
```
where did the n * n come from ? it s not in the doco

lindsayad replied:
Fixed in doco

GiudGiud commented on framework/include/hdgkernels/HDGKernel.h:
```diff
@@ -0,0 +1,173 @@
+
+  /// The auxiliary system which holds the primal variables. We will update the primal variable
+  /// degrees of freedom based off the Lagrange multiplier increment computed by the linear solve
+  /// (during the Newton iteration)
```
```suggestion
  /// This update is performed on every nonlinear (for example Newton) iteration.
```

GiudGiud commented on test/tests/hdgkernels/diffusion/tests:
```diff
@@ -0,0 +1,77 @@
+    []
+  []
+  [rt_mms]
+    requirement = 'The system shall show first order convergence in the L2 norm of the scalar field when using constants for the scalar basis and first order L2 Raviart-Thomas basis functions for the vector field in a hybridized mixed method on'
```
do we expect 1st order with RT?

lindsayad replied:
Yes the scalar field is approximated with constants, so k + 1 = 1 convergence is what we expect

GiudGiud commented on framework/src/hdgkernels/DiffusionHDGAssemblyHelper.C:
```diff
@@ -0,0 +1,403 @@
+                                                 const QBase & qrule)
+{
+  for (const auto qp : make_range(qrule.n_points()))
+    for (const auto i : make_range(_vector_n_dofs))
```
I would guess flipping these two loops would be faster. seems like it would be vectorizable over Qps even

but only a guess

GiudGiud commented on framework/include/hdgkernels/HDGData.h:
```diff
@@ -0,0 +1,45 @@
+   * The primal and Lagrange multiplier variables that contribute to the HDG discretization
+   */
+  virtual std::set<const MooseVariableBase *> variables() const = 0;
+
```
```suggestion

// Note that both primal equations use the same residual vector and matrix, using different indexes (offsets) to store both residuals / matrices
```

GiudGiud commented on framework/include/hdgkernels/HDGData.h:
```diff
@@ -0,0 +1,45 @@
+
+  /// Matrix data structures for on-diagonal coupling
+  EigenMatrix _PrimalMat, _LMMat;
+  /// Vector data structures
```
is this always just for the residual ? if so then maybe specify it? If not then can we list what goes in there or would it not help

GiudGiud commented on modules/navier_stokes/doc/content/source/hdgkernels/NavierStokesHDGKernel.md:
```diff
@@ -0,0 +1,21 @@
+
+where $\rho$ is the density, $\mu$ is the dynamic viscosity and $p$ is the pressure. This class uses
+the hybridization laid out in [!citep](nguyen2011implicit).
+
```
would be good to mention the integration by parts since it matters for implementing more BCs

GiudGiud replied:
also the weak forms for the equations
this page can then stand-in for the hdgns.md we will eventually want to link to from the main table of all NS discretizations we have

GiudGiud replied:
umm I guess the reference to nguyen is good enough for the equations

GiudGiud replied:
umm I guess the reference to nguyen is good enough for the equations

GiudGiud replied:
it's not open access but I can find it here
https://acdl.mit.edu/AIAA2010_362.pdf

GiudGiud reviewed (approved):
looks great!

GiudGiud commented on framework/doc/content/source/hdgkernels/DiffusionHDGKernel.md:
```diff
@@ -0,0 +1,29 @@
+# DiffusionHDGKernel
+
+This class implements the same physics as [MatDiffusion.md] but with a
```
```suggestion
This class implements the same physics as [MatDiffusion.md]$^*$ but with a
```

GiudGiud commented on framework/doc/content/source/hdgkernels/DiffusionHDGKernel.md:
```diff
@@ -0,0 +1,29 @@
+variable $u$, and the third is for the trace or Lagrange multiplier variable
+$\hat{u}$.
+
+!syntax parameters /HDGKernels/DiffusionHDGKernel
```
```suggestion

!alert note
Unlike MatDiffusion, the dependence of the diffusion coefficient on nonlinear variables is not captured in the construction of the Jacobian.

!syntax parameters /HDGKernels/DiffusionHDGKernel
```

lindsayad committed:
Address elements of Guillaume review

- More doxygen
- Documentation corrections
- Note on LDG vs. IP methods

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

lindsayad committed:
Swap qp and ij loop ordering where easy

lindsayad committed:
Use functor for pressure mms function

Allows removal of TransientInterface::time

grmnptr reviewed (approved):
Nice new capability!
